File: fund-system/server/param-connectivity.js
// ════════════════════════════════════════════════════════════
// PARAM CONNECTIVITY
//
// A control is only real if the scanner reads the value at runtime. The
// store holds 189 rows across 10 scanners; Block 3 established that only
// two scanners demonstrably read anything. A panel exposing all of them
// without that distinction would look authoritative and mostly do
// nothing.
//
// Status is derived from EXECUTED node data, never from workflow source.
// A param present in source but absent from executed data is exactly the
// case this exists to catch.
//
//   CONNECTED       the stored VALUE appears in run data.
//                   params_source:'server' alone is NOT sufficient -- that
//                   object carries no parameter values, so it evidences a
//                   fetch, not use. That is the mean_reversion failure mode.
//   NO_RUNTIME_READ zero appearances across a MEANINGFUL sample (>=30)
//                   BY A SCANNER THAT ACTUALLY RAN its trading path.
//   INDETERMINATE   enough executions, but none reached the trading path --
//                   halted, or out of window. Absence here is evidence
//                   about the scanner's state, not about the parameter.
//                   indices read CONNECTED=13 one day and 0 the next purely
//                   because its breaker short-circuited before the config
//                   node emitted anything. Both readings were honest; the
//                   badge was not.
//   UNPROVEN        sample thinner than that, or the name appears without
//                   the stored value. Absence of evidence is not evidence
//                   of absence and must not render as a negative.
//   NOT_WIRED       no active workflow declares this scanner.
//
// Scanning is expensive, so the result is cached with an as_of stamp and
// refreshed on demand.
// ════════════════════════════════════════════════════════════
import { execFile } from 'child_process';
import path from 'path';
import { fileURLToPath } from 'url';
import pg from 'pg';

const HERE = path.dirname(fileURLToPath(import.meta.url));

// Lazy, like param-store.js:56. ESM imports are evaluated before the
// importing module's body runs, so reading process.env at module scope
// would sample it before dotenv.config() has populated it.
let pool = null;
function getPool() {
  if (pool) return pool;
  if (!process.env.DATABASE_URL) return null;
  pool = new pg.Pool({ connectionString: process.env.DATABASE_URL, max: 2 });
  return pool;
}

export const MEANINGFUL_SAMPLE = 30;
const CACHE_TTL_MS = 30 * 60_000;
let cache = { data: null, at: 0 };

function runReader(pairs, timeoutMs = 90_000) {
  return new Promise((resolve, reject) => {
    const child = execFile('python3', [path.join(HERE, 'param-connectivity.py')],
      { timeout: timeoutMs, maxBuffer: 32 * 1024 * 1024 },
      (err, stdout, stderr) => {
        if (err) return reject(new Error(`param-connectivity failed: ${err.message} ${String(stderr).slice(0, 200)}`));
        try { resolve(JSON.parse(stdout)); }
        catch (e) { reject(new Error(`param-connectivity returned non-JSON: ${e.message}`)); }
      });
    child.stdin.end(JSON.stringify({ pairs }));
  });
}

export async function computeParamConnectivity({ force = false } = {}) {
  if (!force && cache.data && Date.now() - cache.at < CACHE_TTL_MS) return cache.data;
  const pg = getPool();
  if (!pg) throw new Error('Postgres is required for param connectivity');

  const { rows } = await pg.query(
    `SELECT scanner, param, value FROM scanner_params WHERE variant='champion' ORDER BY scanner, param`);
  const res = await runReader(rows.map(r => ({ scanner: r.scanner, param: r.param, value: r.value })));

  const summary = { CONNECTED: 0, NO_RUNTIME_READ: 0, INDETERMINATE: 0, UNPROVEN: 0, NOT_WIRED: 0 };
  const byScanner = {};
  for (const p of res.pairs) {
    summary[p.status] = (summary[p.status] || 0) + 1;
    const s = (byScanner[p.scanner] ||= { CONNECTED: 0, NO_RUNTIME_READ: 0, INDETERMINATE: 0, UNPROVEN: 0, NOT_WIRED: 0, total: 0 });
    s.execution_state = p.execution_state ?? null;
    s.execution_note = p.execution_note ?? null;
    s[p.status]++; s.total++;
  }

  const data = {
    as_of: new Date().toISOString(),
    scan_window: res.scan_window,
    meaningful_sample: MEANINGFUL_SAMPLE,
    total_rows: res.pairs.length,
    summary,
    by_scanner: byScanner,
    // The one number that says how much of the control panel is real.
    real_control_rows: summary.CONNECTED,
    pairs: res.pairs
  };
  cache = { data, at: Date.now() };
  return data;
}

export function connectivityCacheAge() {
  return cache.data ? Math.round((Date.now() - cache.at) / 1000) : null;
}

File: fund-system/server/param-connectivity.py
"""Which (scanner, param) pairs does the fleet ACTUALLY read at runtime?

Reads n8n execution_data and decodes the flatted format. Deliberately
does NOT read workflow source: a param present in source but absent from
executed data is exactly the case this exists to catch.

Reads {"pairs":[{"scanner":..,"param":..,"value":..}]} on stdin, emits
evidence per pair on stdout. READ-ONLY (mode=ro URI).
"""
import sqlite3, json, os, re, sys

DB = os.environ.get("N8N_SQLITE_DB", "/var/lib/docker/volumes/n8n_data/_data/database.sqlite")
LIMIT = int(os.environ.get("PARAM_CONN_SCAN", "60"))
TAG_RE = r"""(?:SCANNER\s*=\s*|["']scanner["']\s*:\s*)['"]([a-z_]+)['"]"""

# A parameter can only be observed if the execution actually reached the
# code that reads it. Detected on RESPONSE-side field names the scanners
# emit, never on request URLs -- URLs are built from variables and never
# appear literally in the flatted blob.
TRADING_RE = re.compile(
    r"signal_fired|rejection_stage|skip_reason|reservation_id|risk_gate|trade_type", re.I)
# A scanner stopped by its own breaker still executes; it just returns
# before the config node emits anything. That is not evidence about a
# parameter.
HALT_RE = re.compile(
    r"SYSTEM_BREAKER|CONSECUTIVE_LOSS_HALT|MAX_ORDERS_PER_DAY|MAX_DAILY_LOSS_R|kill_switch|BREAKER", re.I)


def main():
    req = json.load(sys.stdin)
    pairs = req.get("pairs", [])
    con = sqlite3.connect(f"file:{DB}?mode=ro", uri=True)
    con.row_factory = sqlite3.Row

    # scanner -> active workflow id, self-declared by the workflow itself
    active = {}
    for r in con.execute("SELECT id,nodes FROM workflow_entity WHERE active=1"):
        found = {}
        for m in re.finditer(TAG_RE, r["nodes"] or ""):
            found[m.group(1)] = found.get(m.group(1), 0) + 1
        if found:
            active.setdefault(max(found, key=found.get), r["id"])

    # one pass per scanner: pull the blobs once, test every param against them
    by_scanner = {}
    for p in pairs:
        by_scanner.setdefault(p["scanner"], []).append(p)

    out = []
    for scanner, plist in by_scanner.items():
        wid = active.get(scanner)
        if not wid:
            for p in plist:
                out.append({**p, "status": "NOT_WIRED", "executions_examined": 0,
                            "executions_with_param": 0, "trading_path_executions": 0,
                            "halted_executions": 0, "execution_state": "NOT_WIRED",
                            "execution_note": "no active workflow declares this scanner",
                            "observed_value": None, "last_observed_at": None,
                            "reason": "no active workflow declares this scanner"})
            continue

        rows = con.execute(
            "SELECT id, startedAt FROM execution_entity WHERE workflowId=? ORDER BY startedAt DESC LIMIT ?",
            (wid, LIMIT)).fetchall()
        blobs = []
        for r in rows:
            d = con.execute("SELECT data FROM execution_data WHERE executionId=?", (r["id"],)).fetchone()
            if d and d["data"]:
                blobs.append((r["startedAt"], d["data"]))

        # What was this scanner DOING over the window? A NO_RUNTIME_READ on
        # a scanner that never reached its trading path says nothing about
        # the parameter -- it says the scanner was halted or out of hours.
        trading = sum(1 for _, b in blobs if TRADING_RE.search(b))
        halted = sum(1 for _, b in blobs if HALT_RE.search(b))
        if not blobs:
            exec_state, exec_note = "NO_EXECUTIONS", "no retained executions in the window"
        elif trading == 0 and halted > 0:
            exec_state, exec_note = "HALTED", f"{halted} of {len(blobs)} executions hit a breaker or halt; none reached the trading path"
        elif trading == 0:
            exec_state, exec_note = "NO_TRADING_PATH", f"all {len(blobs)} executions were heartbeat-only; the trading path never ran (out of window, or gated upstream)"
        elif halted > trading:
            exec_state, exec_note = "MOSTLY_HALTED", f"{halted} halted vs {trading} trading-path executions"
        else:
            exec_state, exec_note = "RUNNING", f"{trading} of {len(blobs)} executions reached the trading path"

        for p in plist:
            name = p["param"]
            # A bare substring test manufactures hits: MAX_HEAT is a prefix
            # of MAX_HEAT_PCT, MAX_POSITIONS of MAX_POSITIONS_PER_X. Require
            # the name to end at a non-word character or a false CONNECTED
            # is only a naming coincidence away.
            edge = r'(?![A-Za-z0-9_])'
            present = re.compile(re.escape(name) + edge)
            val_a = re.compile(re.escape(name) + edge + r'"\s*,\s*"?([0-9.eE+-]{1,24})"?')
            val_b = re.compile(re.escape(name) + edge + r'["\s:,]+([0-9.eE+-]{1,24})')
            hits, seen, last_at = 0, set(), None
            for started, b in blobs:
                if not present.search(b):
                    continue
                hits += 1
                if last_at is None:
                    last_at = started
                # flatted stores values in a string table; capture what sits
                # near the key in either form
                for m in val_a.finditer(b):
                    seen.add(m.group(1))
                for m in val_b.finditer(b):
                    seen.add(m.group(1))

            stored = str(p.get("value", ""))
            matched = any(v == stored or (
                _num(v) is not None and _num(stored) is not None and _num(v) == _num(stored)
            ) for v in seen)

            if hits > 0 and matched:
                status, reason = "CONNECTED", f"value {stored} observed in {hits} of {len(blobs)} executions"
            elif hits > 0:
                # the name appears but never with the stored value
                status = "UNPROVEN"
                reason = (f"param name seen in {hits} executions but stored value {stored} "
                          f"not observed (values seen: {sorted(seen)[:4]})")
            elif len(blobs) >= 30 and trading > 0:
                status = "NO_RUNTIME_READ"
                reason = (f"zero appearances across {len(blobs)} executions, of which {trading} "
                          f"reached the trading path — the scanner ran and did not read this value")
            elif len(blobs) >= 30:
                # Enough executions, but NONE reached the code that could
                # read a parameter. Absence here is not evidence of
                # absence -- it is evidence about the scanner's state.
                status = "INDETERMINATE"
                reason = (f"not observed, but no execution reached the trading path in this window "
                          f"({exec_note}) — this says nothing about the parameter")
            else:
                status = "UNPROVEN"
                reason = f"only {len(blobs)} executions retained — too thin to conclude"

            out.append({**p, "status": status, "executions_examined": len(blobs),
                        "executions_with_param": hits,
                        "trading_path_executions": trading,
                        "halted_executions": halted,
                        "execution_state": exec_state, "execution_note": exec_note,
                        "observed_value": sorted(seen)[0] if seen else None,
                        "last_observed_at": last_at, "reason": reason})

    json.dump({"pairs": out, "scan_window": LIMIT}, sys.stdout)


def _num(v):
    try:
        return float(v)
    except Exception:
        return None


if __name__ == "__main__":
    main()
